### ISSUE TITLE
Comparison by md5 and large objects

### ISSUE THREAD

**Issue by xgiovio** [ncw/rclone#203]
Hello, i haven't tried yet but i was wondering how rclone manage comparison by md5 on large objects.
Usually files on openstack are >0 bytes if they are full files, 0 bytes if they are manifests of a file divided in segments, a folder or a simply 0 byte file.

Now for files > 0 and 0 byte files we have an etag metadata that stores md5 values. So using rclone it would be possible to compare file using md5.
For folders rclone should simply skip these files checking metadata content-type application/folder
but
for large objects?
Openstack stores in the manifest an etag as the md5 of contatenation of all md5(etag) segments.
It's impossible then to have a comparison for that file without download it first.

In addition rclone comparison supports only -c for size and checksum, no options for mtimes and size, size-only for size only.
But if i want to sync/copy a folder that contains both large and normal files i should have an intermediate option : size-only for large files (comparing the content-length metadata), size + checksum for normal files.

The software should list all files, divide the list into files >0 and 0.
For >0 do the normal size + cheksum comparison (checking size and etag)
For 0 files check the metadatas: if the metadatas contain the manifest it should compare only the content-length, if it's a folder skip it, if a simple 0 file do the file size + checksum comparison.

What do you think?

**Comment by ncw** [comment]
For large objects in Swift that were uploaded with multipart, then the md5sum is useless and rclone will ignore it.  Rclone only segments objects bigger than 5 GB BTW.

If the MD5SUM is missing then rclone assumes that the md5sums match.

A normal sync uses (size, mod time) to see if files are identical.  If size is the same, but mod time is different then it uses md5sum to check if the files are really the same.  If they are then it resets the mod time.  If the md5sum isn't present then it is equivalent to it matching.

If you do a `--checksum` sync, then rclone will compare (size, md5sum) for files, but if the md5sum is missing it will assume it matches, so it will just be comparing (size).

So, I think you'll find rclone works exactly as you suggest already :-)

Checkout the [Equal function](https://github.com/ncw/rclone/blob/master/fs/operations.go#L93) if you are interested in the details.

**Comment by xgiovio** [comment]
hello, thank you for the answer. I have verified your steps but rclone is bugged. That's why.
#Test1 without large objects

On local folder i have differents iso files. Each is less than 5 gb. On swift side, i have the same files but with different mtimes and sizes = 0 bytes.
_________________________________________
First run: I've tried to copy the files from local to hubic with -n -v
from the log i see:
file1.iso have different mtimes. So it starts a md5 check. Spent 2 minutes to create md5sum and say:
returning empty md5sum for manifest file. after this : size of src and dst objects identical.
The files on swift have the mdsum in the etag, they are not large. Why it shouldn't be avaible?

WTF!
Point1. It should check the size and not the mtimes at the beginning. It should compare local filesystem file size with swift content-length and for redundancy the size reported from listing folder.(here rclone should say that they already are different and not equal)
Point2. If you (rclone) haven't tried to compare the size why are you running an md5 check? Useless.
I've started a copy without any options so it should compare first the size and if it is equal then the mtimes check. If the mtimes check fail then the md5 check. If the md5 isn't avaible then the file are the same. But on first run it should compare size and if them are different it should stop and mark it for copy!

The situation is the same for the other isos. 
_________________________________________
Second run: I've tried to copy the files from local to hubic with -n -v -c
Now rclone says in no time the sizes differ and mark it for copy. Good, it doesn't run the md5 check.

Tested also with little identical files, it checks the size and after the md5 check. Right
____________________________________________

I'll report different tests on large obejcts but i have different questions:
1) How rclone check the size : content-legth or directory listing
2) How rclone identify large objects ? The only method is check the presenza of manifest metadata. It shouldn't check the file size because an user can upload a file with chunks of 10mb for example.

**Comment by ncw** [comment]
> On swift side, i have the same files but with different mtimes and sizes = 0 byte.

`sizes == 0` imples that they are manifest files doesn't it? Which explains the above.

> The files on swift have the mdsum in the etag, they are not large. Why it shouldn't be avaible?

rclone thinks they are manifest files, the Etag is the Etag of the manifest file, and therefore not useful.

It is unfortunate rclone does the md5sum of the local file first before finding the remote has no md5sum, but that is the way it goes sometimes!

> Point1. It should check the size and not the mtimes at the beginning. It should compare local filesystem file size with swift content-length and for redundancy the size reported from listing folder.(here rclone should say that they are already different and not equal)

That is exactly what rclone does.  It checks the size first. rclone thinks the sizes are the same, hence the md5sum check.

> Second run: I've tried to copy the files from local to hubic with -n -v -c
> Now rclone says in no time the sizes differ and mark it for copy. Good, it doesn't run the md5 check.

That is inconsistent with the first run where rclone thought the files were the same length...

> 1) How does rclone check the size : content-legth or directory listing

It can be either.

> 2) How does rclone identify large objects ? The only method is checking the presence of manifest metadata. It shouldn't check the file size because an user can upload a file with chunks of 10mb for example.

Only by looking for `X-Object-Manifest`

----

After a bit of testing, I see the problem.

When rclone updates the modification time, it is removing the `X-Object-Manifest` header which explains the inconsistent runs above.

Here is my reproduction

```
mkdir test2a
dd if=/dev/zero of=test2a/100k bs=1k count=100
rclone -v --swift-chunk-size=16k copy test2a swift:test2a
touch test2a/100k
rclone -v --swift-chunk-size=16k copy test2a swift:test2a
# This last one uploads the file again saying sizes differ
rclone -v --swift-chunk-size=16k copy test2a swift:test2a
```

It should be easy to fix

**Comment by xgiovio** [comment]
when rclone finds a 0 byte file should first check if it is a large object.
>sizes == 0 imples that they are manifest files doesn't it?
is uncorrect.

Often there are files with 0 byte with title only. Example: edited_by_xgiovio.txt
is a txt file used as marker, it's a 0 byte object but not a manifest.

For this reason the first test doesn't compare the iso size with the 0 byte file.

When rclone find a 0 byte file, it should 
1) check if both local and remote are 0 bytes. If yes stop
2) If they are different (local bigger than 0) it should check the remote metadata. If it dosn't have a manifest the file is wrong and than the file should be overwritten. If there is a manifest, it should compare the content-length. If equal all is ok otherwise it should overwrite.

**Comment by xgiovio** [comment]
>When rclone updates the modification time, it is removing the X-Object-Manifest header which explains the inconsistent runs above.

This is a very big and serious bug. If the manifest has gone you can't relink it if the folder of segments is on in the thousands of folder of container_segment. It can ruins terabytes of uploads!

**Comment by ncw** [comment]
> when rclone finds a 0 byte file should first check if it is a large object.

That is exactly what it does!

If you tell me what OS and bits (32/64) I'll send you a binary to test the fix.

**Comment by xgiovio** [comment]
Win64

**Comment by ncw** [comment]
Try [this](http://pub.rclone.org/windows-amd64/rclone.exe).

I also put in the not yet released hubic remote (do `rclone config` and you'll see)

Let me know how you get on

**Comment by xgiovio** [comment]
yes it seems to work.

For hubic consider that the access_token has about 1 day expire time. So rclone shoul grab another token for long sessions with refresh token.

I'll open another thread for another discussion

**Comment by ncw** [comment]
The Swift auth token normally expires and there is a mechanism in swift for fetching a new one when it does.  The hubic remote implements that, though I haven't actually tested it yet!

**Comment by ncw** [comment]
This fix has gone live in v1.25

Please test it and if you find any problems then open a new issue.

Thanks

Nick
